File: browser-demo/extension-standalone/src/core/policy.ts
import type { IntentProposal } from './intent';
import type { SemanticEvent } from './semantic_event';
import type { CompiledWorld } from '../world/manifest_schema';
import { evaluatePolicyFromWorld } from './world_runtime';

export type PolicyDecision = 'allow' | 'deny' | 'ask' | 'simulate';

export interface PolicyRuleDescriptor {
  rule_id: string;
  rule_description: string;
  decision: PolicyDecision;
}

export interface PolicyResult {
  decision: PolicyDecision;
  rule_hit: string;        // kept for backward compat — always equals rule_id
  rule_id: string;
  rule_description: string;
  explanation: string;
}

// Exported static rule table — used by world_state.ts to render the world model
export const POLICY_RULES: PolicyRuleDescriptor[] = [
  {
    rule_id: 'always_allow_summarize',
    rule_description: 'Summarizing page content is a safe read-only operation always permitted regardless of source trust or taint.',
    decision: 'allow'
  },
  {
    rule_id: 'allow_read_only_extraction',
    rule_description: 'Extracting links and action items are non-mutating read operations permitted from any source.',
    decision: 'allow'
  },
  {
    rule_id: 'deny_export_for_tainted_content',
    rule_description: 'Exporting a summary is blocked when the page contains hidden or tainted content to prevent data exfiltration of injected payloads.',
    decision: 'deny'
  },
  {
    rule_id: 'ask_before_memory_write_from_untrusted',
    rule_description: 'Writing to memory from an untrusted source requires explicit user approval to prevent prompt-injection-driven memory poisoning.',
    decision: 'ask'
  },
  {
    rule_id: 'ask_before_export_from_untrusted',
    rule_description: 'Exporting content from an untrusted source requires explicit user approval since the content may have been manipulated.',
    decision: 'ask'
  },
  {
    rule_id: 'fallback_simulate_unknown',
    rule_description: 'Any intent not matched by a specific rule is simulated rather than executed, ensuring unknown operations cannot cause unintended side effects.',
    decision: 'simulate'
  }
];

/**
 * Evaluate a policy decision.
 *
 * When `compiledWorld` is supplied the decision is delegated to the dynamic
 * world runtime, which reads rules from the active world manifest.
 *
 * When `compiledWorld` is absent the legacy hardcoded rules below are used.
 * This fallback ensures backward-compatibility during the transition period
 * and on first-run before the balanced_world preset is activated.
 */
export function evaluatePolicy(
  event: SemanticEvent,
  intent: IntentProposal,
  compiledWorld?: CompiledWorld
): PolicyResult {
  if (compiledWorld) {
    return evaluatePolicyFromWorld(compiledWorld, event, intent);
  }
  return evaluatePolicyLegacy(event, intent);
}

function evaluatePolicyLegacy(event: SemanticEvent, intent: IntentProposal): PolicyResult {
  // Rule 1: Always allow summarization
  if (intent.intent_type === 'summarize_page') {
    return {
      decision: 'allow',
      rule_hit: 'always_allow_summarize',
      rule_id: 'always_allow_summarize',
      rule_description: POLICY_RULES[0].rule_description,
      explanation: `Summarizing "${event.url}" is a safe read-only operation; trust level and taint do not restrict this.`
    };
  }

  // Rule 2: Allow read-only extraction
  if (intent.intent_type === 'extract_links' || intent.intent_type === 'extract_action_items') {
    return {
      decision: 'allow',
      rule_hit: 'allow_read_only_extraction',
      rule_id: 'allow_read_only_extraction',
      rule_description: POLICY_RULES[1].rule_description,
      explanation: `Extraction from "${event.url}" is non-mutating; no side effects are possible.`
    };
  }

  // Rule 3: Deny export if tainted (checked before trust level)
  if (event.taint && intent.intent_type === 'export_summary') {
    return {
      decision: 'deny',
      rule_hit: 'deny_export_for_tainted_content',
      rule_id: 'deny_export_for_tainted_content',
      rule_description: POLICY_RULES[2].rule_description,
      explanation: `Page at "${event.url}" is tainted (hidden content detected); exporting could leak injected data.`
    };
  }

  // Rule 4: Ask before memory write from untrusted source
  if (event.trust_level === 'untrusted' && intent.intent_type === 'save_memory') {
    return {
      decision: 'ask',
      rule_hit: 'ask_before_memory_write_from_untrusted',
      rule_id: 'ask_before_memory_write_from_untrusted',
      rule_description: POLICY_RULES[3].rule_description,
      explanation: `"${event.url}" is untrusted; writing to memory without confirmation risks poisoning future agent behaviour.`
    };
  }

  // Rule 5: Ask before export from untrusted source (only reached if taint is false)
  if (event.trust_level === 'untrusted' && intent.intent_type === 'export_summary') {
    return {
      decision: 'ask',
      rule_hit: 'ask_before_export_from_untrusted',
      rule_id: 'ask_before_export_from_untrusted',
      rule_description: POLICY_RULES[4].rule_description,
      explanation: `"${event.url}" is untrusted; exporting its content requires explicit confirmation.`
    };
  }

  // Rule 6: Fallback — simulate unknown intents (no side effects)
  return {
    decision: 'simulate',
    rule_hit: 'fallback_simulate_unknown',
    rule_id: 'fallback_simulate_unknown',
    rule_description: POLICY_RULES[5].rule_description,
    explanation: `Intent "${intent.intent_type}" matched no explicit rule; defaulting to simulate to prevent unintended side effects.`
  };
}

File: browser-demo/service/app/policy.py
"""
Deterministic world policy engine.

Rules are pure Python — no LLM, no probabilistic logic.
First matching rule wins (ordered list).

Trust assignment
  web_page            → untrusted
  extension_ui        → trusted
  manual_user_input   → trusted
  <anything else>     → untrusted

Taint assignment
  trust == "untrusted"        → tainted
  hidden_content_detected     → tainted (always, regardless of trust)

Intent rules (evaluated in order, first match wins)
  save_memory   + untrusted                    → ask
  export_summary + tainted                     → deny
  summarize_page                               → allow
  extract_links                                → allow
  extract_action_items                         → allow
  save_memory   + trusted                      → allow
  export_summary + not tainted                 → allow
  <default>                                    → deny
"""
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


# ---------------------------------------------------------------------------
# Trust
# ---------------------------------------------------------------------------

_TRUST_MAP: dict[str, str] = {
    "web_page": "untrusted",
    "extension_ui": "trusted",
    "manual_user_input": "trusted",
}


def assign_trust(source_type: str) -> str:
    return _TRUST_MAP.get(source_type, "untrusted")


# ---------------------------------------------------------------------------
# Taint
# ---------------------------------------------------------------------------

def assign_taint(trust: str, hidden_content_detected: bool) -> bool:
    """Taint is monotonically joined — once tainted, always tainted."""
    return trust == "untrusted" or hidden_content_detected


# ---------------------------------------------------------------------------
# Available actions (shown in the UI before any specific intent is fired)
# ---------------------------------------------------------------------------

_ALL_INTENTS = [
    "summarize_page",
    "extract_links",
    "extract_action_items",
    "save_memory",
    "export_summary",
]


def available_actions(trust: str, taint: bool) -> list[str]:
    """
    Return the list of intents the user can trigger.

    All intents are always surfaced so the demo can show the deny/ask
    outcomes. (Hiding the button would hide the interesting result.)
    """
    return list(_ALL_INTENTS)


# ---------------------------------------------------------------------------
# Evaluation result
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class PolicyDecision:
    decision: str       # allow | deny | ask | simulate
    rule_hit: str
    reason: str


# ---------------------------------------------------------------------------
# Rule table
# ---------------------------------------------------------------------------

@dataclass(frozen=True)
class _Rule:
    intent: Optional[str]           # None matches any
    trust: Optional[str]            # None matches any
    taint: Optional[bool]           # None matches any
    hidden: Optional[bool]          # None matches any
    decision: str
    rule_hit: str
    reason: str

    def matches(
        self,
        intent: str,
        trust: str,
        taint: bool,
        hidden: bool,
    ) -> bool:
        if self.intent is not None and self.intent != intent:
            return False
        if self.trust is not None and self.trust != trust:
            return False
        if self.taint is not None and self.taint != taint:
            return False
        if self.hidden is not None and self.hidden != hidden:
            return False
        return True


_RULES: list[_Rule] = [
    # Malicious override attempt: hidden instructions + save_memory → deny
    _Rule(
        intent="save_memory",
        trust=None,
        taint=None,
        hidden=True,
        decision="deny",
        rule_hit="RULE-SM-HIDDEN",
        reason="Memory write blocked: hidden content detected in page source",
    ),
    # save_memory from untrusted source → ask user
    _Rule(
        intent="save_memory",
        trust="untrusted",
        taint=None,
        hidden=None,
        decision="ask",
        rule_hit="RULE-SM-UNTRUSTED",
        reason="Memory write from untrusted page requires explicit user approval",
    ),
    # export_summary when content is tainted → deny
    _Rule(
        intent="export_summary",
        trust=None,
        taint=True,
        hidden=None,
        decision="deny",
        rule_hit="RULE-EX-TAINTED",
        reason="Export blocked: content carries taint from untrusted or hidden source",
    ),
    # summarize_page → always allow (read-only, no side effects)
    _Rule(
        intent="summarize_page",
        trust=None,
        taint=None,
        hidden=None,
        decision="allow",
        rule_hit="RULE-SUMMARIZE-ALLOW",
        reason="Page summarization is a read-only operation; permitted from any source",
    ),
    # extract_links → always allow
    _Rule(
        intent="extract_links",
        trust=None,
        taint=None,
        hidden=None,
        decision="allow",
        rule_hit="RULE-LINKS-ALLOW",
        reason="Link extraction is read-only; permitted from any source",
    ),
    # extract_action_items → always allow
    _Rule(
        intent="extract_action_items",
        trust=None,
        taint=None,
        hidden=None,
        decision="allow",
        rule_hit="RULE-ACTION-ITEMS-ALLOW",
        reason="Action item extraction is read-only; permitted from any source",
    ),
    # save_memory from trusted source → allow
    _Rule(
        intent="save_memory",
        trust="trusted",
        taint=None,
        hidden=None,
        decision="allow",
        rule_hit="RULE-SM-TRUSTED",
        reason="Memory write from trusted extension UI is permitted",
    ),
    # export_summary from clean (non-tainted) content → allow
    _Rule(
        intent="export_summary",
        trust=None,
        taint=False,
        hidden=None,
        decision="allow",
        rule_hit="RULE-EX-CLEAN",
        reason="Export from clean, non-tainted content is permitted",
    ),
    # Default: deny anything not explicitly allowed
    _Rule(
        intent=None,
        trust=None,
        taint=None,
        hidden=None,
        decision="deny",
        rule_hit="RULE-DEFAULT-DENY",
        reason="No matching allow rule; denied by default",
    ),
]


def evaluate(
    intent_type: str,
    trust: str,
    taint: bool,
    hidden_content_detected: bool,
) -> PolicyDecision:
    """
    Evaluate an intent against the world policy.

    Processes rules in declaration order; returns the first match.
    This function is deterministic: identical inputs always produce
    identical outputs.
    """
    for rule in _RULES:
        if rule.matches(intent_type, trust, taint, hidden_content_detected):
            return PolicyDecision(
                decision=rule.decision,
                rule_hit=rule.rule_hit,
                reason=rule.reason,
            )
    # Should be unreachable because the default-deny rule matches everything.
    return PolicyDecision(
        decision="deny",
        rule_hit="RULE-FALLBACK-DENY",
        reason="Fallback deny — reached end of rule table",
    )


# ---------------------------------------------------------------------------
# World description (for /world/current endpoint)
# ---------------------------------------------------------------------------

WORLD_TRUST_DEFAULTS = {
    "web_page": "untrusted",
    "extension_ui": "trusted",
    "manual_user_input": "trusted",
}

WORLD_TAINT_DEFAULTS = {
    "untrusted_source": "tainted",
    "hidden_content": "tainted (always)",
    "trusted_source": "clean",
}

WORLD_INTENT_POLICY_SUMMARY = [
    {"intent": "summarize_page",       "condition": "any",                   "decision": "allow"},
    {"intent": "extract_links",        "condition": "any",                   "decision": "allow"},
    {"intent": "extract_action_items", "condition": "any",                   "decision": "allow"},
    {"intent": "save_memory",          "condition": "trusted source",        "decision": "allow"},
    {"intent": "save_memory",          "condition": "untrusted, no hidden",  "decision": "ask"},
    {"intent": "save_memory",          "condition": "hidden content",        "decision": "deny"},
    {"intent": "export_summary",       "condition": "tainted content",       "decision": "deny"},
    {"intent": "export_summary",       "condition": "clean content",         "decision": "allow"},
    {"intent": "<default>",            "condition": "any",                   "decision": "deny"},
]
